src/OpenTelemetry/Metrics/MetricPoint.cs

```diff
@@ -247,12 +247,18 @@ internal void Update(double number)
                 case AggregationType.Histogram:
                     {
                         int i;
-                        for (i = 0; i < this.histogramBuckets.ExplicitBounds.Length; i++)
+                        if (double.IsNaN(number))
                         {
-                            // Upper bound is inclusive
-                            if (number <= this.histogramBuckets.ExplicitBounds[i])
+                            i = this.histogramBuckets.ExplicitBounds.Length;
+                        }
+                        else
+                        {
+                            // If `number` is not a histogram bound, the result will be negative
+                            // The bitwise complement of the returned value is the insertion index
+                            i = Array.BinarySearch(this.histogramBuckets.ExplicitBounds, number);
```

Comment:
For array with small number of bounds, this extra invocation could turn out to be slower.

We need to have a perf benchmark to show the numbers.

Refer to the prom-client and see how low/high buckets are handled differently:
* https://github.com/prom-client-net/prom-client/blob/master/src/Prometheus.Client/HistogramImpl/HistogramHighBucketsStore.cs
* https://github.com/prom-client-net/prom-client/blob/master/src/Prometheus.Client/HistogramImpl/HistogramLowBucketsStore.cs

Reply:
I'll write a self-contained benchmark that compares the existing linear search implementation with the binary search one and run it with a range of boundary counts.

I agree that linear is likely faster for a small number of boundaries, but eventually the binary search will outperform it. Then I can a boundary number where anything less uses linear, and anything more uses binary.

I'll generate the random numbers so that each index gets an equal amount of finds.
Ex: If the sorted array is [ 20, 30, 40 ] then the random numbers will be generated like `random.next(10, 50)`
An array of all the results from performing 10,000 checks should looks similar to:

| Value | Count |
|-------|--------|
| 0 | 2500 |
| 1 | 2500 |
| 2 | 2500 |
| 3 | 2500 |